jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java

```diff
@@ -122,17 +123,57 @@ public boolean isPassivating()
     @Override
     public Set<String> doGetExpired( Set<String> candidates )
     {
-        if (candidates == null || candidates.isEmpty())
-        {
-            return Collections.emptySet();
-        }
+        Set<String> expired = new HashSet<>();
         
         long now = System.currentTimeMillis();
-        return candidates.stream().filter( candidate -> {
+        expired.addAll( sessionDataMap.values( mapEntry ->  {
+            SessionData sessionData = (SessionData) mapEntry.getValue();
+            if (_context.getWorkerName().equals(sessionData.getLastNode()))
+            {
+                //we are its manager, add it to the expired set if it is expired now
+                if ((sessionData.getExpiry() > 0 ) && sessionData.getExpiry() <= now)
+                {
+                    if (LOG.isDebugEnabled())
+                    {
+                        LOG.debug( "Session {} managed by {} is expired", sessionData.getId(), _context.getWorkerName() );
+                    }
+                    return true;
+                }
+            }
+            else
+            {
+                //if we are not the session's manager, only expire it iff:
+                // this is our first expiryCheck and the session expired a long time ago
+                //or
+                //the session expired at least one graceperiod ago
+                if (_lastExpiryCheckTime <=0)
+                {
+                    if ((sessionData.getExpiry() > 0 ) && sessionData.getExpiry() < (now - (1000L * (3 * _gracePeriodSec))))
+                    {
+                        return true;
+                    }
+                }
+                else
+                {
+                    if ((sessionData.getExpiry() > 0 ) && sessionData.getExpiry() < (now - (1000L * _gracePeriodSec)))
+                    {
+                        return true;
+                    }
+                }
+            }
+            return false;
+        } ).stream().map( sessionData -> sessionData.getId() ).collect( Collectors.toSet()));
+
+        //check candidates that were not found to be expired, definitely they no longer exist and they should be expired
+
+        expired.addAll( candidates.stream().filter( candidate -> {
             
             if (LOG.isDebugEnabled())
                 LOG.debug( "Checking expiry for candidate {}", candidate );
-            
+            if(expired.contains( candidate ))
+            {
+                return false;
+            }
             try
             {
                 SessionData sd = load(candidate);
```

Comment:
Is it possible to test for the existance of a key instead?

Reply:
yup good idea might be good optimisation